serde_arrow/src/internal/serialization_ng/list_builder.rs

```diff
@@ -0,0 +1,187 @@
+use serde::{ser::SerializeSeq, Serialize};
+
+use crate::{
+    internal::{common::MutableOffsetBuffer, error::fail},
+    Error, Result,
+};
+
+use super::{array_builder::ArrayBuilder, not_implemented::NotImplemented};
+
+#[derive(Debug, Clone)]
+
+pub struct ListBuilder {
+    pub offsets: MutableOffsetBuffer<i32>,
+    pub element: Box<ArrayBuilder>,
+}
+
+impl ListBuilder {
+    pub fn new(element: ArrayBuilder) -> Self {
+        Self {
+            offsets: Default::default(),
+            element: Box::new(element),
+        }
+    }
+}
+
+impl<'a> serde::Serializer for &'a mut ListBuilder {
+    type Error = Error;
+    type Ok = ();
+
+    type SerializeMap = NotImplemented;
+    type SerializeSeq = &'a mut ListBuilder;
+    type SerializeStruct = NotImplemented;
+    type SerializeTuple = NotImplemented;
+    type SerializeStructVariant = NotImplemented;
+    type SerializeTupleStruct = NotImplemented;
+    type SerializeTupleVariant = NotImplemented;
+
+    fn serialize_unit(self) -> Result<()> {
+        fail!("Serializer::serialize_unit is not implemented")
```

Comment:
Amazing that you're writing it so quickly! Thanks! 
In case you may be interested, I've also written [this](https://crates.io/crates/serde_serializer_quick_unsupported) tiny crate specifically for this scenario where you have most variants give "not implemented": https://crates.io/crates/serde_serializer_quick_unsupported
It greatly eases the process and reduces the boilerplate of writing such serializers.

Reply:
Thanks. That would've been helpful. I now went with a slightly different approach: 

- Use a custom trait [`SimpleSerializer`](https://github.com/chmp/serde_arrow/blob/edea994ccdedf11d393f58be7a76a48e13b8f432/serde_arrow/src/internal/serialization_ng/utils.rs#L32) (the name is debatable :))
- Add a [blanket impl of `Serailizer`](https://github.com/chmp/serde_arrow/blob/edea994ccdedf11d393f58be7a76a48e13b8f432/serde_arrow/src/internal/serialization_ng/utils.rs#L306) for `SimpleSerializer`. The `Mut` wrapper is required to satisfy Rust's orphan rules

Reply:
Ah but it looks like that prevents from using the associated types, which causes to still need to have some form of runtime state machine through functions like `serialize_struct_variant_start`, whereas otherwise this could be a fully type-based state machine, where at every point the type that is currently executing its method determines the current state.
I'm guessing the impact of this depends on the extent to which:
- The state is complex (haven't looked yet) 
- The optimizer decides to inline, enabling it to actually get rid of the state updates through const propagation and noalias guarantees. 

This is probably a small impact...

Reply:
You were right again :) 

For structs, I opted to keep the current machinary at the cost of having "next" and "seen" fields in the builder. The cleaner option would to have a separate object that tracks this temporary correctness info. TBH., I am okay with this small code smell, as there are bigger fish to fry (like the clearning up the deserialization impl :D). Performancewise this choice does not make a difference, as the switches between builders is handled by points where serde gives you an object to serialize (e.g., `serialize_field<T: Serialize>(&mut self, field: &str, value: &T)`).

The only case where it really makes a performance difference is for enums in Rust / unions in Arrow. There I followed the serde interface an return the builder for the variant:

```rust
pub trait SimpleSerializer: Sized {
  /.../

  fn serialize_struct_variant_start<'this>(
        &'this mut self,
        name: &'static str,
        variant_index: u32,
        variant: &'static str,
        len: usize,
    ) -> Result<&'this mut ArrayBuilder>;
    
   fn serialize_tuple_variant_start<'this>(
        &'this mut self,
        name: &'static str,
        variant_index: u32,
        variant: &'static str,
        len: usize,
    ) -> Result<&'this mut ArrayBuilder>;
}